policyeval/broker.go

```diff
@@ -0,0 +1,347 @@
+package policyeval
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/helper/uuid"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
+)
+
+// Broker stores, dedups and control access to policy evaluation requests.
+type Broker struct {
+	logger hclog.Logger
+
+	l sync.RWMutex
+
+	// nackTimeout is the time required for a dequeued eval to be ack'd.
+	nackTimeout time.Duration
+
+	// deliveryLimit is the number of times an eval can be dequeued before
+	// being considered as failed.
+	deliveryLimit int
+
+	// pendingEvals holds evaluations that are ready to be picked for
+	// further evaluation.
+	// Each key represents a different queue and the value is a heap holding
+	// the evals in order of priority and create time.
+	pendingEvals map[string]PendingEvaluations
+
+	// enqueuedEvals and enqueuedPolicies are indexes to quickly find if an eval
+	// or a policy is waiting be evaluated.
+	enqueuedEvals    map[string]int
+	enqueuedPolicies map[string]string
+
+	// unack tracks evaluations that have not been ack'd yet.
+	unack map[string]*unackEval
+
+	// waiting tracks Dequeue requests that are blocked waiting for work.
+	waiting map[string]chan struct{}
```

Comment:
it might be nice to document the invariants between these fields
* every named queue should be represented as a key in `pendingEval` and `waiting`
* an eval should either be present in both `enqueuedEvals` and `unack`, or present in `pendingEvals`

Reply:
That's  a good idea and something I struggled with while reading the Nomad code 😅